packages/server-wallet/src/onchain-service/onchain-service.ts

```diff
@@ -0,0 +1,297 @@
+   *
+   * @notice This may not be strictly necessary, but is useful for testing
+   */
+  public detachAllHandlers(assetHolderAddr: Address, event?: ContractEvent): void {
+    const record = this.assetHolders.get(assetHolderAddr);
+    if (!record) {
+      throw new Error(`Could not find asset holder with service`);
+    }
+
+    // Only detach handlers from one event if specified
+    if (event) {
+      record.evts[event].detach();
+      return;
+    }
+
+    // Remove all handlers (user and default)
+    Object.values(record.evts).map(evt => {
+      evt.detach();
+    });
+  }
+
+  /**
+   * Adds a channel to watch events and submit transactions for
+   *
+   * @param channelId Unique channel identifier to register chain listeners for
+   * @param assetHolders Onchain addresses of the channel chain context
+   * @returns an empty promise that will resolve once the channel is registered
+   */
+  public async registerChannel(channelId: Bytes32, assetHolders: Address[]): Promise<void> {
+    if (!this.channelWallet) {
+      throw new OnchainServiceError(OnchainServiceError.reasons.noChannelWallet);
+    }
+
+    // Add the channel if it has not been stored
+    if (!this.store.has(channelId)) {
+      this.store.set(channelId, {});
+    }
+
+    // Create and store new contracts if we don't have record of
+    // required assetHolders
+    await Promise.all(
+      assetHolders.map(async assetHolderAddr => {
+        // If the contract has already been registered, continue
+        if (this.assetHolders.has(assetHolderAddr)) {
+          return;
+        }
+
+        // Get asset holder information from chain
+        const info = await getAssetHolderInformation(assetHolderAddr, this.provider);
+        this._registerAssetHolderCallbacks(info);
+      })
+    );
+  }
+
+  public async submitTransaction(
+    channelId: Bytes32,
+    tx: MinimalTransaction
+  ): Promise<providers.TransactionResponse> {
+    if (!this.channelWallet) {
+      throw new OnchainServiceError(OnchainServiceError.reasons.noChannelWallet);
+    }
+
+    // If the channel is not registered, do not send transactions
+    if (!this.store.has(channelId)) {
+      throw new OnchainServiceError(OnchainServiceError.reasons.notRegistered);
+    }
+
+    return this.transactionSubmissionService.submitTransaction(tx, {
+      maxSendAttempts: this.attempts,
+    });
+  }
+
+  private _createDepositEvt(): Evt<DepositedEvent & {block: number}> {
+    // Create the evt instances for the contract
+    const depositEvt = Evt.create<DepositedEvent & {block: number}>();
+
+    // Setup deposit evt so it will emit properly formatted event IFF
+    // - belongs to a registered channel
+    // - is for a subsequent deposit event
+    // And execute channel wallet and storage callbacks (wallet callback
+    // executed IFF a channel wallet has been attached to the service)
+    depositEvt
+      .pipe(e => {
+        return this.store.has(e.destination);
```

Comment:
Maybe transform the value into `this.store.get(e.destination)` after do you don't need to call it in subsequently piped functions